externals/llvm-external-projects/torch-mlir-dialects/include/torch-mlir-dialects/Dialect/Tcp/IR/TcpOps.td

```diff
@@ -0,0 +1,65 @@
+#ifndef TORCH_MLIR_DIALECT_TCP_OPS
+#define TORCH_MLIR_DIALECT_TCP_OPS
+
+include "torch-mlir-dialects/Dialect/Tcp/IR/TcpOpsBase.td"
+
+include "mlir/IR/OpBase.td"
+include "mlir/Interfaces/SideEffectInterfaces.td"
+
+def Tcp_TanhOp : Tcp_UnaryOp<"tanh", [NoSideEffect]> {
+  let summary = "Computes tanh of input, elementwise";
+
+  let description = [{
+    Elementwise unary op to compute tanh.
+  }];
+
+  let arguments = (ins
+    Tcp_Tensor:$in
+  );
+
+  let results = (outs
+    Tcp_Tensor:$out
+  );
+
+  let assemblyFormat = "$in attr-dict `:` type($in) `->` type($out)";
+}
+
+def Tcp_AddOp : Tcp_BinaryOp<"add", [NoSideEffect]> {
+  let summary = "Computes elementwise addition";
+
+  let description = [{
+    Elementwise addition.
+  }];
+
+  let arguments = (ins
+    Tcp_Tensor:$in1,
+    Tcp_Tensor:$in2
+  );
+
+  let results = (outs
+    Tcp_Tensor:$out
+  );
+
+  let assemblyFormat = "$in1 `,` $in2 attr-dict `:` type($in1) `,` type($in2) `->` type($out)";
+}
+
+def Tcp_MatmulOp : Tcp_Op<"matmul", [NoSideEffect]> {
```

Comment:
I think it would be good to implement verification + type inference for this op right away. This will force us to face interesting questions, in addition to the 2D/3D item which is already discussed. E.g.:
  1) Do we allow result element type to be different from operand element type? I think we should, if we want to support quantization well.
  2) If yes, what do we do with type inference? Do we: a) say that matmul doesn't support it, b) add something like `preferred_element_type` attribute to enable type inference, c) infer result element type equal to operand element type and then work around via isCompatibleReturnTypes?

Reply:
+1,

Also, why not have separate matmul and batch_matmul ops? My experience is that with these "variant" ops it's much easier to `isa<BatchMatmulOp>` than `if (getOperand(0).cast<RankedTensorType>().getRank() == 3)`

Reply:
I'm working on a document to summarize the different design choices for Matmul (2 ops vs 1 op, 3D and 2D cases, etc.). So, I'm removing it from this PR. Will send a separate PR for matmul later.

Reply:
> why not have separate matmul and batch_matmul ops?

I'm working on a document to summarize the different design choices for Matmul (2 ops vs 1 op, 3D and 2D cases, etc.). So, I'm removing matmul op from this PR. Will send a separate PR for matmul later.